extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java

```diff
@@ -85,6 +114,168 @@ void registerSecurityInterceptors(BuildProducer<InterceptorBindingRegistrarBuild
+            MethodCreator methodCreator,
+            Map.Entry<MethodInfo, AnnotationInstance> methodEntry) {
+        try {
+            MethodInfo method = methodEntry.getKey();
+            ResultHandle aClass = methodCreator.load(method.declaringClass().name().toString());
+            ResultHandle methodName = methodCreator.load(method.name());
+            ResultHandle params = paramTypes(methodCreator, method.parameters());
+
+            AnnotationInstance instance = methodEntry.getValue();
+            ResultHandle securityAnnotation = methodCreator.load(instance.name().toString());
+
+            ResultHandle annotationParameters = annotationValues(methodCreator, instance);
+
+            Method registerAnnotation = SecurityCheckStorageBuilder.class.getDeclaredMethod("registerAnnotation",
+                    String.class, String.class, String[].class, String.class, String[].class);
+            methodCreator.invokeVirtualMethod(MethodDescriptor.ofMethod(registerAnnotation), checkStorage,
+                    aClass, methodName, params, securityAnnotation, annotationParameters);
+        } catch (NoSuchMethodException e) {
+            throw new IllegalStateException("registerAnnotation method not found on on SecurityCheckStorage", e);
+        }
+    }
+
+    private ResultHandle annotationValues(MethodCreator methodCreator, AnnotationInstance instance) {
+        AnnotationValue value = instance.value();
+        if (value != null && value.asStringArray() != null) {
+            String[] values = value.asStringArray();
+            ResultHandle result = methodCreator.newArray(String.class, methodCreator.load(values.length));
+            int i = 0;
+            for (String val : values) {
+                methodCreator.writeArrayValue(result, i++, methodCreator.load(val));
+            }
+            return result;
+        }
+        return methodCreator.loadNull();
+    }
+
+    private ResultHandle paramTypes(MethodCreator ctor, List<Type> parameters) {
+        ResultHandle result = ctor.newArray(String.class, ctor.load(parameters.size()));
+
+        for (int i = 0; i < parameters.size(); i++) {
+            ctor.writeArrayValue(result, i, ctor.load(parameters.get(i).toString()));
+        }
+
+        return result;
+    }
+
+    private Map<MethodInfo, AnnotationInstance> gatherSecurityAnnotationsByLooping(ApplicationIndexBuildItem indexBuildItem,
+            BeanRegistrar.RegistrationContext registrationContext,
+            boolean loopThroughIndex) {
+        Set<DotName> securityAnnotations = SecurityAnnotationsRegistrar.SECURITY_BINDINGS.keySet();
+        AnnotationStore annotationStore = registrationContext.get(BuildExtension.Key.ANNOTATION_STORE);
+        Set<ClassInfo> classesWithSecurity = new HashSet<>();
+
+        Index index = indexBuildItem.getIndex();
+        if (loopThroughIndex) { // only needed when synthetic security annotations are added
```

Comment:
I don't really like this approach with two different code paths. IMHO a better approach would be to DenyAllBuildItem  be AdditionalSecuredClassesBuildIem, and have it just direct provide the class names of the JAX-RS endpoints, so no iteration is required and there is only one code path.

Reply:
How would you handle the other denyAll case though, I mean the one that applies to all CDI beans? There isn't a way to get all beans as far as I know.

Reply:
Look up @DenyAll from the annotation index

Reply:
Basically for the other case where we want to deny it only applies if there is another security annotation on the CdI bean. This means the existing logic in the other branch should discover them.

Reply:
That won't work for the `quarkus.securitu.deny-unannotated` option.
It won't work because that option means that we need to implicitly add DenyAll to all beans that don't have any security annotation, therefore there is nothing to look up in the index.

Reply:
Now that raises the question of whether this config option makes sense at all. I personally don't understand it, but I am assuming it's been discussed before and there was good reason to add it.

Reply:
for CDI beans, the synthetic deny all only applies to beans that have some security annotations, so that shouldn't be a problem.
Testing once again. I had problems with build steps cycle but that clearly shouldn't be a problem since `DenyAllBuildItem` is passed from step consuming a RestEasy deployment...

Reply:
It adds DenyAll to methods of a bean that have at least one security annotation:

```
class Foo {
@RolesAllowed("foo")
 void doStuff1() {}

void doOtherStuff() {}

}
```
In this case it will mark doOtherStuff as DenyAll. Basically it is a failsafe against forgetting to add security metadata to a secured bean, you need to explicitly use @PermitAll to state your intentions.

Reply:
@michalszynkiewicz if you can get rid of the loop for CDI beans, then the JAX-RS case could be easily handled like @stuartwdouglas proposes.
I am however sceptical that the CDI case can be handled without the loop, but I'd like it very much if it can

Reply:
Aha! Know I see what I misunderstood... I thought that DenyAll applies to ALL beans, ever without any security annotation.
So yeah I agree this should be easy and there is no need for the loop